lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp

```diff
@@ -512,20 +464,22 @@ assignMemoryLayouts(llvm::SmallVector<std::tuple<Operation *, int, Operation *>>
     }
     loadToInfo[op] = loadInfo;
   }
-
-  return loadToInfo;
 }
 
-static llvm::MapVector<Operation *, LoadInfo>
-scheduleLoads(scf::ForOp forOp, tt::CoarseSchedule &schedule,
-              DenseSet<Operation *> &rootUsers, int numStages) {
+// assignMemoryLayouts, loadToInfo
+static tt::CoarseSchedule::Cluster
+getLoopSchedule(scf::ForOp forOp, tt::CoarseSchedule &schedule,
+                /*DenseSet<Operation *> &rootUsers,*/ int numStages,
+                llvm::MapVector<Operation *, LoadInfo> &loadToInfo) {
   ModuleOp moduleOp = forOp->getParentOfType<ModuleOp>();
   tt::ModuleAxisInfoAnalysis axisInfoAnalysis(moduleOp);
+  tt::CoarseSchedule::Cluster afterPrologue = schedule.clusters.begin();
 
   // Get all loads that are (transitively) used by dot ops and their distance
   // to the dot op.
   llvm::SmallVector<std::tuple<Operation *, int, Operation *>>
-      loadOpToIndLevelAndUse = loadOpsToIndirectionLevelAndUse(forOp);
+      loadOpToIndLevelAndUse =
+          mlir::triton::loadOpsToIndirectionLevelAndUse(forOp);
```

Comment:
This part still needs some rework. loadOpToIndLevelAndUse is used by assignMemoryLayouts, and also distToUse for the load ops. I don't think the indirection level is needed. But the list of loads and uses for each load are needed.

Reply:
Right, it seems like this is needed only by assignMemoryLayouts, and for two reasons:
1. Check if the load feeds directly to the dot
2. Check if load feeds into another load, to ensure that if the second load is not getting pipelined, the first one that feeds into it also shouldn't be

Re. 1 - This can be probably replaced by some simpler check
Re. 2 - I feel like this kind of decisions should be part of a Scheduling pass. The SWP should be relatively dumb I feel, just discovering dependencies, and never override decision that comes from the Scheduling pass whether to pipeline something or not. These kind of decisions will be very schedule-dependent (in this case we need to worry about it, because schedule handles dependent loads).
One possible problem is that we are actually dropping the loads based on the memory layout (because of the workaround for a bug), but I still feel that we need to make this decision earlier

Reply:
@pawelszczerbuk @ThomasRaoux
Do you think we should move part of assignMemoryLayouts to loopScheduling, especially this part:
```
    } else if (auto loadOp = dyn_cast<tt::LoadOp>(use)) {
      // The use of this loadOp is another loadOp. If the use is not in the
      // loadsToPipeline already, it means that the use is not valid for
      // pipelining for some reason. We should skip this loadOp, too. Note that
      // we have an assumption that distAndUse.second (i.e. the use of this
      // loadOp) has already be processed in a previous loop iteration. This
      // assumption is held by how loadOpsToIndirectionLevelAndUse recursively
      // collects loadOpToIndLevelAndUse using DFS.
      if (loadToInfo.count(loadOp) == 0) {
        continue;
      }
    }
```
We can do a bit more filtering at loopScheduling.
As to the fields, I am not sure which fields should be set in loopScheduling.
```
struct LoadInfo {
  // Layout of the data in the shared memory.
  ttg::SharedEncodingAttr sharedEncoding = nullptr;
  // Blocked encoding is used for loads not used by the dot.
  ttg::BlockedEncodingAttr blockedEncoding = nullptr;
  bool loadIsMMAV3 = false;
  int distToUse = 0;
  bool usedByDot = false;
};
```

Also assignMemoryLayouts currently looks at loadOpToIndLevelAndUse, but it only looks at the first use, since for the 2nd use, loadToInfo.count(op) will be true and it will be ignored.
For distToUse, we are looking at all uses, but the last use will win:
```
  for (auto [loadOp, _, use] : loadOpToIndLevelAndUse) {
    if (loadToInfo.count(loadOp) == 0)
      continue;
    loadToInfo[loadOp].distToUse = schedule[use].first - schedule[loadOp].first;
  }
```
Is it okay for SWP to figure out the direct dot use on its own for each loadOp? i.e a simpler version of loadOpToIndLevelAndUse